docs/RFCS/20200720_modern_passwords.md

```diff
@@ -0,0 +1,590 @@
+
+- some user accounts are already created in v20.1, with method `bcrypt-crdb`
+- the CockroachDB version is upgraded, offering a new pw storage method
+- new user accounts get created with the new method
+
+Because it's not possible to convert password details stored with one method
+to another method, we must support multiple methods side-by-side.
+
+We do this by adding a new column `format` to `system.users`:
+
+- its DEFAULT value is `bcrypt-crdb`. The default makes it
+  backward-compatible with all the rows already stored in the table.
+- the new value `scram-sha-256` indicates that the `hashedPassword`
+  field stores SCRAM-SHA-256 parameters.
+
+During authentication, the authn code verifies that the stored values
+in `system.users` are compatible with the authn method requested by
+the HBA configuration.
+
+This achieves [Goal 1 outlined in the summary](#summary).
+
+## Version gating
+
+In a mixed version X/Y config, where version X does not yet support the
+new storage method, we must prevent the creation of user accounts
+using the new method.
+
+Otherwise, it would be possible for version X nodes to mistakenly
+use the hashed details of the new method using the `bcrypt-crdb`
+algorithm, resulting in a potentially insecure situation.
+
+For this, we introduce a cluster version to gate usage of the new method.
+
+## Default storage method selection
+
+When a client provides a password in plaintext using `CREATE ROLE WITH
+PASSWORD`, the server must choose a storage method to compute
+`hashedPassword` in `system.users`.
+
+This is done by introducing a cluster setting
+`server.user_login.password_encryption`. This `password_encryption`
+setting has the same role as `password_encryption` in PostgreSQL: it
+determines the storage method to use for all new accounts, when
+the client configures an initial password in cleartext.
+
+In CockroachDB, the only two allowed values are `bcrypt-crdb` and
+`scram-sha-256`.
+
+We introduce a cluster setting as opposed to hard-coding the algorithm
+so as to facilitate tests and the generation of `system.users` tables
+compatible with previous versions of CockroachDB.
+
+## Storage method selection in WITH PASSWORD
+
+We want to make it possible to create password-based accounts without
+passing the cleartext password to the server.
+
+This is used e.g. in the CC management console to make the server
+blind to the cleartext password.
+
+For this we extend the SQL dialect as follows:
+
+- (baseline, already implemented:) `CREATE USER/ROLE WITH PASSWORD
+  'xxxx'`: the server understands `xxxx` as a cleartext password.
+
+  The server computes `system.users.hashedPassword` using the storage
+  method configured server-side via the setting
+  `server.user_login.password_encryption`.
+
+  The current value of `server.user_login.password_encryption` is stored
+  in the (new) `system.users.format` column.
+
+- `CREATE USER/ROLE WITH PASSWORD 'xxxx' ENCRYPTION 'yyyy'`: the
```

Comment:
The [postgres CREATE ROLE docs](https://www.postgresql.org/docs/12/sql-createrole.html) say

>  If the presented password string is already in MD5-encrypted or SCRAM-encrypted format, then it is stored as-is ...

Should we do that too so we don't need to add syntax unsupported by postgres? You discuss below the bcrypt prefixes, which perhaps answers this.